### PR TITLE
Use CoreTiming for MemoryWatcher.

### PR DESCRIPTION
**Pull request by spxtr**
It might be worth moving some code from `Core.cpp` into `MemoryWatcher.cpp`. Let me know.

---
This change is [Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/3791)

### CODE REVIEW

**Comment by JosJuice** [comment]
Is there any particular reason for this change? Determinism? Simplicity?

---

Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[Source/Core/Core/Core.cpp, line 118 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3791#-KFwmYkr1qtFZ3tzUCaW:-KFwmYkr1qtFZ3tzUCaX:2003381092) ([raw file](https://github.com/dolphin-emu/dolphin/blob/fa6d73f9994bbfa03e394e7897924f35debb71a3/Source/Core/Core/Core.cpp#L118)):*
Please use longer names. `fps` is fine to use as an acronym (though it's inaccurate in this case – MemoryWatcher callbacks are not frames), but `mw` should be memory_watcher, and I don't even know what `et` is supposed to be.

---

*[Source/Core/Core/Core.cpp, line 123 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3791#-KFwnFk8zEM6U4_r_YZl:-KFwnFk9iYMrDT7J_rQZ:-1744694777) ([raw file](https://github.com/dolphin-emu/dolphin/blob/fa6d73f9994bbfa03e394e7897924f35debb71a3/Source/Core/Core/Core.cpp#L123)):*
You might want to subtract `cyclesLate` from the next callback's delay.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3791#-:-KFwmVCcdcr6Q7nd46cr:-792237322)*

File: Source/Core/Core/Core.cpp [hunk 1]
```diff
@@ -114,6 +114,14 @@ static bool s_is_throttler_temp_disabled = false;
 
 #ifdef USE_MEMORYWATCHER
 static std::unique_ptr<MemoryWatcher> s_memory_watcher;
+static int et_MW;
+static const int s_mw_fps = 600;
+
+static void MWCallback(u64 userdata, s64 cyclesLate)
+{
+	s_memory_watcher->Step();
+	CoreTiming::ScheduleEvent(SystemTimers::GetTicksPerSecond() / s_mw_fps, et_MW);
```
Comment (degasus) [inline Source/Core/Core/Core.cpp]: "- cyclesLate", so timing will be less affected by eg jit block merging.

File: Source/Core/Core/Core.cpp [hunk 2]
```diff
@@ -362,6 +371,8 @@ static void CpuThread()
 
 #ifdef USE_MEMORYWATCHER
```
Comment (degasus) [inline Source/Core/Core/Core.cpp]: MemoryWatcher::Init();

File: Source/Core/Core/Core.cpp [hunk 3]
```diff
@@ -289,6 +297,7 @@ void Stop()  // - Hammertime!
 #endif
 
 #ifdef USE_MEMORYWATCHER
```
Comment (degasus) [inline Source/Core/Core/Core.cpp]: MemoryWatcher::Shutdown();

File: Source/Core/Core/MemoryWatcher.cpp [hunk 4]
```diff
@@ -85,30 +80,29 @@ std::string MemoryWatcher::ComposeMessage(const std::string& line, u32 value)
 	return message_stream.str();
 }
 
-void MemoryWatcher::WatcherThread()
+void MemoryWatcher::Step()
 {
-	while (m_running)
+	if (!m_running)
```
Comment (degasus) [inline Source/Core/Core/MemoryWatcher.cpp]: We're on the CPU thread now, so this flag isn't needed any more.

**Comment by degasus** [comment]
In my opinion, there is no need to ask why something should be deterministic. As long as there is no good reason not to be so, everything should be determinnistic ;)

So I like this change. A dummy thread is dropped, a race condition is dropped (you had accessed the GC memory in a unsafe way). But yeah, moving more code into MemoryWatcher.cpp would be nice.

**Comment by JosJuice** [comment]
Indeed, it's better to have this deterministic. I was just asking because it wasn't clear at first that determinism was the reason for this change.

**Comment by spxtr** [comment]
Determinism is nice, but the real reason is that I want to run it at roughly 5x speed, and polling at a fixed time interval wasn't good enough for that.

Thanks for the comments, PTAL.

---

Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions.

---

*[Source/Core/Core/Core.cpp, line 118 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3791#-KFwmYkr1qtFZ3tzUCaW:-KG0DJAUt5cMQDiLXJyE:990168306) ([raw file](https://github.com/dolphin-emu/dolphin/blob/fa6d73f9994bbfa03e394e7897924f35debb71a3/Source/Core/Core/Core.cpp#L118)):*
Ok, picked some better names. I took `et` from `SystemTimers.cpp`.

---

*[Source/Core/Core/Core.cpp, line 123 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3791#-KFwf7yT8otOBwj7cuef-r1-123:-KG0DI9azvFQV4aZviY3:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/fa6d73f9994bbfa03e394e7897924f35debb71a3/Source/Core/Core/Core.cpp#L123)):*
Done.

---

*[Source/Core/Core/Core.cpp, line 123 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3791#-KFwnFk8zEM6U4_r_YZl:-KG0DHsL-u2NtnyrUAxq:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/fa6d73f9994bbfa03e394e7897924f35debb71a3/Source/Core/Core/Core.cpp#L123)):*
Done.

---

*[Source/Core/Core/Core.cpp, line 299 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3791#-KFwf7yT8otOBwj7cuef-r1-299:-KG0DH7g8k3f-ZZpUO19:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/fa6d73f9994bbfa03e394e7897924f35debb71a3/Source/Core/Core/Core.cpp#L299)):*
Done.

---

*[Source/Core/Core/Core.cpp, line 372 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3791#-KFwf7yT8otOBwj7cuef-r1-372:-KG0DGYcTv4Ge-QkI_Es:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/fa6d73f9994bbfa03e394e7897924f35debb71a3/Source/Core/Core/Core.cpp#L372)):*
Done.

---

*[Source/Core/Core/MemoryWatcher.cpp, line 85 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3791#-KFwf7yT8otOBwj7cueg-r1-85:-KG0G8SjbAFSeFnCwI4Q:-463360720) ([raw file](https://github.com/dolphin-emu/dolphin/blob/fa6d73f9994bbfa03e394e7897924f35debb71a3/Source/Core/Core/MemoryWatcher.cpp#L85)):*
I still want it because if there is no `Locations.txt` or if the socket doesn't open then I want to skip this. It's certainly possible to get that behavior by only scheduling the event if loading is successful, but this was easier.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3791#-:-KG0BrQCMXav99Yxje45:363905882)*

**Comment by JosJuice** [comment]
Review status: 0 of 3 files reviewed at latest revision, 7 unresolved discussions.

---

*[Source/Core/Core/MemoryWatcher.cpp, line 25 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3791#-KG18VVWIvRy8kEgEHRB:-KG18VVX3eqcGv9vHH8Y:-560570028) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c794cf152a06c67137028a861312d4567951fd8f/Source/Core/Core/MemoryWatcher.cpp#L25)):*
Shouldn't this be `SystemTimers::GetTicksPerSecond() / MW_RATE - cyclesLate`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3791)*

**Comment by Parlane** [comment]
@dolphin-emu-bot rebuild

**Comment by spxtr** [comment]
Review status: 0 of 3 files reviewed at latest revision, 7 unresolved discussions.

---

*[Source/Core/Core/MemoryWatcher.cpp, line 25 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3791#-KG18VVWIvRy8kEgEHRB:-KGAUouylpf7ycVvI6jk:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c794cf152a06c67137028a861312d4567951fd8f/Source/Core/Core/MemoryWatcher.cpp#L25)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3791)*

**Comment by degasus** [comment]
Reviewed 2 of 3 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3791)*

**Comment by JosJuice** [comment]
Please squash the commits.

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3791#-:-KGTZJPlCV3xF_GfRduU:775492945)*

**Comment by spxtr** [comment]
Done.

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3791#-:-KGV_QPECFN4ht2H8m4O:-499146009)*

**Comment by degasus** [comment]
LGTM. IMO fine for 5.0.

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3791#-:-KGW00p2bKfFMKXGTBLR:-1294681228)*

**Comment by vladfi1** [comment]
So it happens that I came up with a different solution to this almost a year ago at https://github.com/vladfi1/dolphin - instead of having a separate thread periodically checking, I had a hook from VideoInterface::EndField that would poll once each frame. Doing this on SSBM gave me no missed frames under a lot of different conditions. Would that be preferable to the current approach?

I also turned  made the memory watcher use zmq (for synchronization) - if we did the same with the named pipes I think that would be enough for windows compatibility.

**Pull request merged**